### PR TITLE
Refactor top level

### PR DESCRIPTION
**Pull request by tgymnich**
I've merged the common boolean flags `fwdMode` and `topLevel` into `DerivativeMode mode` since they were often used in combination to convey the current mode throughout Enzyme.
Combining these flags will clean up some ambiguity.

### CODE REVIEW

File: enzyme/Enzyme/GradientUtils.cpp [hunk 1]
```diff
@@ -2099,10 +2099,10 @@ GradientUtils *GradientUtils::CreateFromClone(
...
 
 DiffeGradientUtils *DiffeGradientUtils::CreateFromClone(
-    EnzymeLogic &Logic, bool topLevel, Function *todiff, TargetLibraryInfo &TLI,
-    TypeAnalysis &TA, DIFFE_TYPE retType, bool diffeReturnArg,
-    const std::vector<DIFFE_TYPE> &constant_args, ReturnType returnValue,
-    Type *additionalArg) {
+    EnzymeLogic &Logic, DerivativeMode mode, Function *todiff,
+    TargetLibraryInfo &TLI, TypeAnalysis &TA, DIFFE_TYPE retType,
+    bool diffeReturnArg, const std::vector<DIFFE_TYPE> &constant_args,
+    ReturnType returnValue, Type *additionalArg) {
```
Comment (wsmoses) [inline enzyme/Enzyme/GradientUtils.cpp:2105]: Add an assert that this is ReverseModeGradient or ReverseModeCombined (and also right now potentially forwardMode, but I think that would be better abstracted into a different subclass of gradientutils).

File: enzyme/Enzyme/GradientUtils.cpp [hunk 2]
```diff
@@ -2037,7 +2037,7 @@ bool GradientUtils::shouldRecompute(const Value *val,
 
 GradientUtils *GradientUtils::CreateFromClone(
     EnzymeLogic &Logic, Function *todiff, TargetLibraryInfo &TLI,
-    TypeAnalysis &TA, DIFFE_TYPE retType,
+    TypeAnalysis &TA, DerivativeMode mode, DIFFE_TYPE retType,
     const std::vector<DIFFE_TYPE> &constant_args, bool returnUsed,
     std::map<AugmentedStruct, int> &returnMapping) {
   assert(!todiff->empty());
```
Comment (wsmoses) [inline enzyme/Enzyme/GradientUtils.cpp:2043]: This shouldn't need to have a mode passed as it always must be ReverseModeAugmented

File: enzyme/Enzyme/EnzymeLogic.cpp [hunk 3]
```diff
@@ -2506,11 +2504,11 @@ void createInvertedTerminator(TypeResults &TR, DiffeGradientUtils *gutils,
...
     Function *todiff, DIFFE_TYPE retType,
     const std::vector<DIFFE_TYPE> &constant_args, TargetLibraryInfo &TLI,
-    TypeAnalysis &TA, bool returnUsed, bool dretPtr, bool topLevel,
+    TypeAnalysis &TA, bool returnUsed, bool dretPtr, DerivativeMode mode,
     llvm::Type *additionalArg, const FnTypeInfo &oldTypeInfo_,
     const std::map<Argument *, bool> _uncacheable_args,
-    const AugmentedReturn *augmenteddata, bool AtomicAdd, bool fwdMode,
-    bool PostOpt, bool omp) {
+    const AugmentedReturn *augmenteddata, bool AtomicAdd, bool PostOpt,
+    bool omp) {
```
Comment (wsmoses) [inline enzyme/Enzyme/EnzymeLogic.cpp:2513]: This should also have an assertion about being reversemodegradient reversemodecombined (and for now forwardmode).

File: enzyme/Enzyme/EnzymeLogic.cpp [hunk 4]
```diff
@@ -21,10 +21,10 @@
 // This file defines two functions CreatePrimalAndGradient and
 // CreateAugmentedPrimal. CreatePrimalAndGradient takes a function, known
 // TypeResults of the calling context, known activity analysis of the
-// arguments and a bool `topLevel`. It creates a corresponding gradient
-// function, computing the forward pass as well if at `topLevel`.
+// arguments. It creates a corresponding gradient
+// function, computing the primal as well.
```
Comment (wsmoses) [inline enzyme/Enzyme/EnzymeLogic.cpp:25]: I'd add if requested

Comment (tgymnich) [reply enzyme/Enzyme/EnzymeLogic.cpp:25]: done

File: enzyme/Enzyme/EnzymeLogic.cpp (hunk 3 above)
Comment (tgymnich) [reply enzyme/Enzyme/EnzymeLogic.cpp:2513]: done

File: enzyme/Enzyme/GradientUtils.cpp (hunk 2 above)
Comment (tgymnich) [reply enzyme/Enzyme/GradientUtils.cpp:2043]: done

File: enzyme/Enzyme/GradientUtils.cpp (hunk 1 above)
Comment (tgymnich) [reply enzyme/Enzyme/GradientUtils.cpp:2105]: done

**Review by wsmoses — APPROVED** [review]

**Pull request merged**